### ISSUE TITLE
关于训练阶段是否传入padding_mask

### ISSUE THREAD

**Issue by kustcl** [jingyaogong/minimind#751]
请问各位大佬，在大佬开源的代码中，在pretrain阶段和sft阶段都没有传入padding_mask,是否会对训练阶段的attention造成污染，影响模型效果。包括推理阶段，处理含有padding的序列，第一次自回归生成是否需要根据padding来确定所选token,而不是取最后一个token预测

**Comment by jingyaogong** [comment]
：不需要、不会污染、0效果影响，看下代码

Pretrain / SFT 数据都是右侧 padding，有效 token 看不到右侧 padding；padding 对应的 label 又会被置为 `-100`，不参与 CE loss

推理是单条输入，没有 padding，取最后一个 token 的 logits 没问题

如果自行做 batch 推理并使用右 padding，那短样本最后一位会是 pad，此时需要改成左 padding，或按 `attention_mask` 找每条样本最后一个有效 token，项目里的 batch rollout 是左 padding，最后位置仍是有效 token。

**Comment by kustcl** [comment]
> ：不需要、不会污染、0效果影响，看下代码
> 
> Pretrain / SFT 数据都是右侧 padding，有效 token 看不到右侧 padding；padding 对应的 label 又会被置为 `-100`，不参与 CE loss
> 
> 推理是单条输入，没有 padding，取最后一个 token 的 logits 没问题
> 
> 如果自行做 batch 推理并使用右 padding，那短样本最后一位会是 pad，此时需要改成左 padding，或按 `attention_mask` 找每条样本最后一个有效 token，项目里的 batch rollout 是左 padding，最后位置仍是有效 token。

感谢大佬的指导，从中学到很多

**Comment by Captainphiora** [comment]
@jingyaogong 作者您好，考虑训练时候的前向传播，如果不传attention_mask的话，padding位置的scores在经过softmax之后就不是0了，这一部分和V相乘之后会流到下一层，这会不会产生影响，虽然padding部分的loss最后不参与计算。
